rita/src/rita_common/peer_listener/mod.rs

```diff
@@ -0,0 +1,364 @@
+impl Message for GetPeers {
+    type Result = Result<Vec<Peer>, Error>;
+}
+
+impl Handler<GetPeers> for PeerListener {
+    type Result = Result<Vec<Peer>, Error>;
+
+    fn handle(&mut self, _: GetPeers, _: &mut Context<Self>) -> Self::Result {
+        Ok(self.peers.clone())
+    }
+}
+
+#[derive(Debug)]
+pub struct ListenInterface {
+    ifname: String,
+    ifidx: u32,
+    multicast_socketaddr: SocketAddrV6,
+    multicast_socket: UdpSocket,
+    linklocal_socket: UdpSocket,
+    linklocal_ip: Ipv6Addr,
+}
+
+impl ListenInterface {
+    pub fn new(ifname: &str) -> Result<ListenInterface, Error> {
+        let port = SETTING.get_network().rita_hello_port;
+        let disc_ip = SETTING.get_network().discovery_ip;
+        trace!("Binding to {:?} for ListenInterface", ifname);
+        // Lookup interface link local ip
+        let link_ip = KI.get_link_local_device_ip(&ifname);
+        if link_ip.is_err() {
+            return Err(PeerListenerError::PeerListenerInterfaceError.into());
+        }
+        let link_ip = link_ip.unwrap();
+        // Lookup interface index
+        let iface_index = if KI.get_iface_index(&ifname).is_ok() {
+            KI.get_iface_index(&ifname).unwrap()
+        } else {
+            0
+        };
+        // Bond to multicast discovery address on each listen port
+        let multicast_socketaddr = SocketAddrV6::new(disc_ip, port.into(), 0, iface_index);
+        let multicast_socket = UdpSocket::bind(multicast_socketaddr)
+            .expect("Failed to bind to peer discovery address!");
+        let res = multicast_socket.join_multicast_v6(&disc_ip, iface_index);
+        trace!("ListenInterface init set multicast v6 with {:?}", res);
+
+        let linklocal_socketaddr = SocketAddrV6::new(link_ip, port.into(), 0, iface_index);
+        let linklocal_socket = UdpSocket::bind(linklocal_socketaddr).expect(&format!(
+            "Failed to bind to link local address {:?} on {:?}",
+            link_ip, ifname
+        ));
+        let res = linklocal_socket.set_nonblocking(true);
+        trace!("ListenInterface init set nonblocking with {:?}", res);
+
+        let res = linklocal_socket.join_multicast_v6(&disc_ip, iface_index);
+        trace!("Set link local multicast v6 with {:?}", res);
+
+        Ok(ListenInterface {
+            ifname: ifname.to_string(),
+            ifidx: iface_index,
+            multicast_socket: multicast_socket,
+            linklocal_socket: linklocal_socket,
+            multicast_socketaddr: multicast_socketaddr,
+            linklocal_ip: link_ip,
+        })
+    }
+}
+
+fn encode_im_here(addr: Ipv6Addr) -> Vec<u8> {
+    let mut buf = Vec::new();
+    buf.put_u32_be(MSG_IM_HERE);
+    buf.put_u16_be(22);
+    let ipaddr_bytes: [u8; 16] = addr.octets();
+    for i in 0..16 {
+        buf.put_u8(ipaddr_bytes[i]);
+    }
+    trace!("Encoded packet {:x?}", buf);
+    buf
+}
+
+fn decode_im_here(buf: &mut Vec<u8>) -> Result<Option<Ipv6Addr>, io::Error> {
```

Comment:
Why do you return a Resul<Option<>> here? I think it would be better to just return a Result<Ipv6Addr, failure::Error>

Reply:
because it's possible everything is ok but we didn't find a proper packet